chips/sifive/src/uart.rs

```diff
@@ -0,0 +1,259 @@
+
+        //            f_clk
+        // f_baud = ---------
+        //           div + 1
+        let divisor = (clock_speed / baud_rate) - 1;
+
+        regs.div.write(div::div.val(divisor));
+    }
+
+    fn enable_tx_interrupt(&self) {
+        let regs = self.registers;
+        regs.ie.modify(interrupt::txwm::SET);
+    }
+
+    fn disable_tx_interrupt(&self) {
+        let regs = self.registers;
+        regs.ie.modify(interrupt::txwm::CLEAR);
+    }
+
+    pub fn handle_interrupt(&self) {
+        let regs = self.registers;
+
+        // Get a copy so we can check each interrupt flag in the register.
+        let pending_interrupts = regs.ip.extract();
+
+        // Determine why an interrupt occurred.
+        if pending_interrupts.is_set(interrupt::txwm) {
+            // Got a TX interrupt which means the number of bytes in the FIFO
+            // has fallen to zero. If there is more to send do that, otherwise
+            // send a callback to the client.
+            if self.len.get() == self.index.get() {
+                // We are done.
+                regs.txctrl.write(txctrl::txen::CLEAR);
+                self.disable_tx_interrupt();
+
+                // Signal client write done
+                self.tx_client.map(|client| {
+                    self.buffer.take().map(|buffer| {
+                        client.transmitted_buffer(buffer, self.len.get(), ReturnCode::SUCCESS);
+                    });
+                });
+            } else {
+                // More to send. Fill the buffer until it is full.
+                self.buffer.map(|buffer| {
+                    for i in self.index.get()..self.len.get() {
+                        // Write the byte from the array to the tx register.
+                        regs.txdata.write(txdata::data.val(buffer[i] as u32));
+                        self.index.set(i + 1);
+                        // Check if the buffer is full
+                        if regs.txdata.is_set(txdata::full) {
+                            // If it is, break and wait for the TX interrupt.
+                            break;
+                        }
+                    }
+                });
+            }
+        }
+    }
+}
+
+impl hil::uart::UartData<'a> for Uart<'a> {}
+impl hil::uart::Uart<'a> for Uart<'a> {}
+
+impl hil::uart::Configure for Uart<'a> {
+    fn configure(&self, params: hil::uart::Parameters) -> ReturnCode {
+        // This chip does not support these features.
+        if params.parity != hil::uart::Parity::None {
+            return ReturnCode::ENOSUPPORT;
+        }
+        if params.hw_flow_control != false {
+            return ReturnCode::ENOSUPPORT;
+        }
+
+        // We can set the baud rate.
+        self.set_baud_rate(params.baud_rate);
+
+        // We need to save the stop bits because it is set in the TX register.
+        self.stop_bits.set(params.stop_bits);
+
+        ReturnCode::SUCCESS
+    }
+}
+
+impl hil::uart::Transmit<'a> for Uart<'a> {
```

Comment:
Any thoughts on the new UART HIL?

Reply:
You know this was just a pretty mechanical translation from the old to the new, so I didn't really interact with the new HIL API. I did like it is more explicit that receive has not been implemented.